### PR TITLE
[UNR-2195] Fix component updates being dropped when waking up an actor

### PR DESCRIPTION
**Pull request by mironec**
#### Description
When an actor wakes up and updates a replicated variable at approximately the same time, it is possible that on the receiving side these updates arrive in reverse order, causing the update to the variable to be dropped. This change fixes this issue by creating a channel on component update - anticipating the wake up.

#### Release note
Should not need one, as it should just go under dormancy fixes.

#### Tests
An intially dormant actor with a repnotified variable - manual.
May need more investigation, which I believe @aleximprobable may do.

#### Documentation
N/A.

#### Primary reviewers
@aleximprobable @m-samiec

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -1123,8 +1124,22 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 	USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(Op.entity_id);
 	if (Channel == nullptr)
 	{
-		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Worker: %s Entity: %d Component: %d - No actor channel for update. This most likely occured due to the component updates that are sent when authority is lost during entity deletion."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
-		return;
+		// If this is a dormant actor and its channel has not been created yet, assume the remove dormant component op comes down later
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can we clear this up a bit. Maybe:
If there is no actor channel as a result of the actor being dormant, then assume the actor is about to become active.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -1123,8 +1124,22 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
...
+		if (const Dormant* DormantComponent = StaticComponentView->GetComponentData<Dormant>(Op.entity_id))
+		{
+			AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Best to be safe, lets not check/crash if the actor is null but output and error message instead

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: It's worth noting that if this means state can change on a client while NetDromancy is set to true. If we are ok with this then we can probably decouple of the concepts of dormancy and receiving updates more.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -221,11 +231,8 @@ void USpatialReceiver::ProcessRemoveComponent(const Worker_RemoveComponentOp& Op
 		if (Op.component_id == SpatialConstants::DORMANT_COMPONENT_ID)
 		{
 			// Receive would normally create channel in ReceiveActor - we need to recreate it here
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can you move this comment to the function

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 4]
```diff
@@ -1123,8 +1130,24 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 	USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(Op.entity_id);
 	if (Channel == nullptr)
 	{
-		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Worker: %s Entity: %d Component: %d - No actor channel for update. This most likely occured due to the component updates that are sent when authority is lost during entity deletion."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
-		return;
+		// If there is no actor channel as a result of the actor being dormant, then assume the actor is about to become active.
+		if (const Dormant* DormantComponent = StaticComponentView->GetComponentData<Dormant>(Op.entity_id))
+		{
+			AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -1123,8 +1130,24 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
...
+			if (Actor != nullptr)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
			if (AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id));)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -1123,8 +1130,24 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
...
+			{
+				RecreateDormantSpatialChannel(Actor, Op.entity_id);
+			}
+			else
+			{
+				UE_LOG(LogSpatialReceiver, Warning, TEXT("Worker: %s Dormant actor (entity: %d) has been deleted on the client but we have received a component update (id: %d) from the server."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
				UE_LOG(LogSpatialReceiver, Warning, TEXT("Worker: %s Dormant actor (entity: %d) has been deleted on this worker but we have received a component update (id: %d) from the server."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 7]
```diff
@@ -163,6 +163,8 @@ class USpatialReceiver : public UObject
 	AActor* TryGetOrCreateActor(SpatialGDK::UnrealMetadata* UnrealMetadata, SpatialGDK::SpawnData* SpawnData);
 	AActor* CreateActor(SpatialGDK::UnrealMetadata* UnrealMetadata, SpatialGDK::SpawnData* SpawnData);
 
+	// Receive would normally create channel in ReceiveActor - this function is used to recreate the channel after waking up a dormant actor
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 8]
```diff
@@ -209,6 +210,15 @@ void USpatialReceiver::RemoveComponentOpsForEntity(Worker_EntityId EntityId)
 	}
 }
 
+void USpatialReceiver::RecreateDormantSpatialChannel(AActor* Actor, Worker_EntityId EntityID)
+{
+	USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(Actor);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	// Receive would normally create channel in ReceiveActor - this function is used to recreate the channel after waking up a dormant actor
	USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(Actor);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -220,12 +231,8 @@ void USpatialReceiver::ProcessRemoveComponent(const Worker_RemoveComponentOp& Op
...
 		if (Op.component_id == SpatialConstants::DORMANT_COMPONENT_ID)
 		{
-			// Receive would normally create channel in ReceiveActor - we need to recreate it here
-			USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(Actor);
-			check(!Channel->bCreatingNewEntity);
-			check(Channel->GetEntityId() == Op.entity_id);
-
-			NetDriver->UnregisterDormantEntityId(Op.entity_id);
+			RecreateDormantSpatialChannel(Actor, Op.entity_id);
+
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -1123,8 +1129,23 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
...
+		// If there is no actor channel as a result of the actor being dormant, then assume the actor is about to become active.
+		if (const Dormant* DormantComponent = StaticComponentView->GetComponentData<Dormant>(Op.entity_id))
+		{
+			if (AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id)))
+			{
+				RecreateDormantSpatialChannel(Actor, Op.entity_id);
+			}
+			else
+			{
+				UE_LOG(LogSpatialReceiver, Warning, TEXT("Worker: %s Dormant actor (entity: %d) has been deleted on this worker but we have received a component update (id: %d) from the server."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
				UE_LOG(LogSpatialReceiver, Warning, TEXT("Worker: %s Dormant actor (entity: %lld) has been deleted on this worker but we have received a component update (id: %d) from the server."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -1123,8 +1129,23 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
...
+			}
+		}
+		else
+		{
+			UE_LOG(LogSpatialReceiver, Log, TEXT("Worker: %s Entity: %d Component: %d - No actor channel for update. This most likely occured due to the component updates that are sent when authority is lost during entity deletion."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
			UE_LOG(LogSpatialReceiver, Log, TEXT("Worker: %s Entity: %lld Component: %d - No actor channel for update. This most likely occured due to the component updates that are sent when authority is lost during entity deletion."), *NetDriver->Connection->GetWorkerId(), Op.entity_id, Op.update.component_id);
```

**Pull request merged**